gslb/nodes/dq_ingestion.go

```diff
@@ -272,6 +300,202 @@ func DequeueIngestion(key string) {
 	case gslbutils.ObjectUpdate:
 		AddUpdateObjOperation(key, cname, ns, objType, objName, sharedQueue, false, SharedAviGSGraphLister())
 	}
+
+}
+
+func OperateOnGSLBHostRule(key string) {
+	_, objType, objName, err := gslbutils.ExtractGSLBHostRuleKey(key)
+	if err != nil {
+		gslbutils.Errf("key: %s, msg: couldn't parse the key for GSLBHostRule: %v", key, err)
+		return
+	}
+	sharedQueue := utils.SharedWorkQueue().GetQueueByName(utils.GraphLayer)
+	AddUpdateGSLBHostRuleOperation(key, objType, objName, sharedQueue, SharedAviGSGraphLister())
+}
+
+func DeleteGSOrGSMembers(aviGSGraph *AviGSObjectGraph, members []AviGSK8sObj, modelName string, agl *AviGSGraphLister,
+	sharedQ *utils.WorkerQueue, key string) {
+	gsName := aviGSGraph.Name
+	for _, m := range members {
+		aviGSGraph.DeleteMember(m.Cluster, m.Namespace, m.Name, m.ObjType)
+	}
+	if len(aviGSGraph.GetUniqueMemberObjs()) == 0 {
+		SharedDeleteGSGraphLister().Save(modelName, aviGSGraph)
+		agl.Delete(modelName)
+	} else {
+		agl.Save(gsName, aviGSGraph)
+	}
+	PublishKeyToRestLayer(utils.ADMIN_NS, gsName, key, sharedQ)
+}
+
+func OperateOnHostRule(key string) {
+	agl := SharedAviGSGraphLister()
+	op, _, cname, _, lfqdn, gfqdn, err := gslbutils.ExtractMultiClusterHostRuleKey(key)
+	if err != nil {
+		gslbutils.Errf("key: %s, msg: couldn't parse the key for HostRule: %v", key, err)
+		return
+	}
+	sharedQ := utils.SharedWorkQueue().GetQueueByName(utils.GraphLayer)
+
+	switch op {
+	case gslbutils.ObjectAdd:
+		fqdnMap := gslbutils.GetFqdnMap()
+		fqdnMap.AddUpdateToFqdnMapping(gfqdn, lfqdn, cname)
+
+		// a Global Fqdn for a Local Fqdn was added, see if we need to delete the GS for local fqdn first
+		var members []AviGSK8sObj
+		modelName := utils.ADMIN_NS + "/" + lfqdn
+		found, aviGS := agl.Get(modelName)
+		if found {
+			aviGSGraph := aviGS.(*AviGSObjectGraph)
+			// an existing GS Graph was found for the local fqdn, delete member(s) or the GS graph.
+			gslbutils.Logf("key: %s, modelName: %s, msg: %s", key, modelName,
+				"GS graph for local fqdn exists, will evaluate the members")
+			members = aviGSGraph.GetGSMembersByCluster(cname)
+			DeleteGSOrGSMembers(aviGSGraph, members, modelName, agl, sharedQ, key)
+		}
+		if len(members) == 0 {
```

Comment:
Should we return here ? what if we process the GSLBHostRule first, before processing the ingress and there is no existing model/member

Reply:
We are returning here. Not sure if I understand the sequence of events. What you are saying is:
- Got an event for `HostRule` for the first time.
- Got an ADD event for an ingress mapped to the `HostRule`.

So, in this case, for the first event, we just return without really doing anything (because the model/member wasn't present). However, we do update the local to global fqdn mapping.
For the second event, we check if the ingress hostname is mapped to a global fqdn, if yes, we create a GS for the global fqdn. Does that make sense?